pandas/core/indexes/multi.py

```diff
@@ -3314,9 +3314,23 @@ def intersection(self, other, sort=False):
         if self.equals(other):
             return self
 
-        self_tuples = self._ndarray_values
-        other_tuples = other._ndarray_values
-        uniq_tuples = set(self_tuples) & set(other_tuples)
+        lvals = self._ndarray_values
+        rvals = other._ndarray_values
+
+        uniq_tuples = None  # flag whether _inner_indexer was succesful
+        if self.is_monotonic and other.is_monotonic:
+            try:
+                uniq_tuples = self._inner_indexer(lvals, rvals)[0]
+                sort = False  # uniq_tuples is already sorted
+            except TypeError:
```

Comment:
im pretty sure ive seen similar code in other Index subclasses.  is there anything we can share?

we only want to catch TypeError from non-comparable types, right?  are there any TypeErrors we might want to not pass on?

Reply:
I could only see the same code in the base Index: https://github.com/pandas-dev/pandas/blob/9d66896cd85fdb08c4fdc945105a39e3d1e796f3/pandas/core/indexes/base.py#L2663-L2667

Yes, the goal is to catch TypeError from non-comparable types. I don't see any other reasons for TypeErrors to be raised but I'm not well-versed enough in pandas implementation to be 100% sure.  `Index.intersection()` does the same thing.

Reply:
thanks for taking a look